modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridCacheDhtPreloadWaitForBackupsTest.java

```diff
@@ -252,6 +272,139 @@ public void testShutdownWithoutBackups() throws Exception {
+        synchronizationMode = CacheWriteSynchronizationMode.FULL_ASYNC;
+        rebalanceMode = CacheRebalanceMode.ASYNC;
+        clientNodes = false;
+        backups = 2;
+
+        startGrids(3);
+
+        grid(0).cluster().active(true);
+
+        grid(0).context().distributedMetastorage().write(
+            DistributedMetaStorageImpl.IGNITE_INTERNAL_KEY_PREFIX + "graceful.shutdown",
+            new HashSet<>(Collections.singleton(grid(0).localNode().id())));
+
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        UUID node1Id = grid(1).localNode().id();
+
+        Thread stopper = new Thread(() -> {
+            grid(1).close();
+            latch.countDown();
+        }, "Stopper");
+
+        stopper.start();
+
+        assertTrue(GridTestUtils.waitForCondition(() -> latch.getCount() == 0, STOP_TIMEOUT_LIMIT));
+
+        HashSet<UUID> expMetastoreContent = new HashSet<>();
+        expMetastoreContent.add(grid(0).localNode().id());
+        expMetastoreContent.add(node1Id);
+
+        assertEquals(
+            expMetastoreContent,
+            grid(0).context().distributedMetastorage().read(GRACEFUL_SHUTDOWN_METASTORE_KEY));
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testThatExcludedNodeListWithinMetastoreCleanedUpAfterUpdatingFullMap() throws Exception {
+        cacheMode = CacheMode.PARTITIONED;
+        atomicityMode = CacheAtomicityMode.ATOMIC;
+        synchronizationMode = CacheWriteSynchronizationMode.FULL_ASYNC;
+        rebalanceMode = CacheRebalanceMode.ASYNC;
+        clientNodes = false;
+        backups = 3;
+
+        startGrids(4);
+
+        grid(0).cluster().active(true);
+
+        grid(0).context().distributedMetastorage().write(
+            DistributedMetaStorageImpl.IGNITE_INTERNAL_KEY_PREFIX + "graceful.shutdown",
+            new HashSet<>(Collections.singleton(grid(0).localNode().id())));
+
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        UUID node1Id = grid(1).localNode().id();
+
+        Thread stopper = new Thread(() -> {
+            grid(1).close();
+            latch.countDown();
+        }, "Stopper");
+
+        stopper.start();
+
+        assertTrue(GridTestUtils.waitForCondition(() -> latch.getCount() == 0, STOP_TIMEOUT_LIMIT));
+
+        HashSet<UUID> expMetastoreContent = new HashSet<>();
+        expMetastoreContent.add(grid(0).localNode().id());
+        expMetastoreContent.add(node1Id);
+
+        assertEquals(
+            expMetastoreContent,
+            grid(0).context().distributedMetastorage().read(GRACEFUL_SHUTDOWN_METASTORE_KEY));
+
+        UUID node2Id = grid(2).localNode().id();
+
+        grid(2).close();
+
+        expMetastoreContent.remove(node1Id);
+        expMetastoreContent.add(node2Id);
+
+        assertEquals(
+            expMetastoreContent,
+            grid(0).context().distributedMetastorage().read(GRACEFUL_SHUTDOWN_METASTORE_KEY));
+    }
+
```

Comment:
I'd extend test coverage with the following tests:
- Customer-like rolling restart case (nodes are restarted one by one, data shouldn't be lost)
- Stress test with simultaneous safe shutdown (all nodes are safely stopped from multiple threads, data shoudn't be lost)
- In-memory / persistent clusters coverage
- Inactive cluster coverage (node stop on inactive cluster shouldn't be delayed)
- Client nodes coverage (metrics is calculated correctly, stop of client nodes shoudn't be delayed)

Reply:
>Customer-like rolling restart case (nodes are restarted one by one, data shouldn't be lost)

org.apache.ignite.internal.processors.cache.distributed.dht.GridCacheDhtPreloadWaitForBackupsTest#testRollingRestartEmulation

>Stress test with simultaneous safe shutdown (all nodes are safely stopped from multiple threads, data shoudn't be lost)

org.apache.ignite.internal.processors.cache.distributed.dht.GridCacheDhtPreloadWaitForBackupsTest#testSimultaneousSafeShutdown

>In-memory / persistent clusters coverage

Already implemented within GridCacheDhtPreloadWaitForBackupsTest and 
GridCacheDhtPreloadWaitForBackupsWithPersistenceTest classes respectively.

>Inactive cluster coverage (node stop on inactive cluster shouldn't be delayed)

org.apache.ignite.internal.processors.cache.distributed.dht.GridCacheDhtPreloadWaitForBackupsTest#testNodeOnInactiveClusterShouldStopImmediately

>Client nodes coverage (metrics is calculated correctly, stop of client nodes shoudn't be delayed)

org.apache.ignite.internal.processors.cache.distributed.dht.GridCacheDhtPreloadWaitForBackupsTest#testClientNodeShouldStopImmediately